index.js

```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+const allowedTokens = ['[', ']', '{}']
+
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
+  return tokens.some(token => allowedTokens.includes(token)) ? tokens : []
+}
+
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+  let token = tokens[0]
+
+  // first token can't be '[', ']', or '{}'
+  if (allowedTokens.includes(token)) isValid = false
+  else if (token === '__proto__' || token === 'constructor') {
+    return `${token} is not allowed as a key name`
+  } else {
+    const bracesEndingCount = tokens.filter(t => t === '{}').length
+
+    // only a single {} special ending is allowed
+    if (bracesEndingCount > 1) isValid = false
+    // {}, if present, must be the last token
+    else if (bracesEndingCount > 0 && tokens[tokens.length - 1] !== '{}') isValid = false
+  }
+
+  const tokensLen = tokens.length
+  let nextToken = tokens[1]
+  let i = 1
+  let hasOpenBracket = false
+
+  while (i < tokensLen) {
+    token = nextToken
+    nextToken = tokens[++i]
+
+    if (token === '__proto__' || token === 'constructor') {
+      return `${token} is not allowed as a key name`
+    }
+
+    // invalid closing bracket
+    if (!hasOpenBracket && token === ']') {
+      isValid = false
+      break
+    }
+    // invalid opening bracket
+    if (hasOpenBracket && token === '[') {
+      isValid = false
+      break
+    }
+
+    hasOpenBracket = token === '[' || (hasOpenBracket && token !== ']')
+    // orphan opening bracket
+    if (hasOpenBracket && nextToken === undefined) isValid = false
+    // {} is not allowed within brackets
+    if (hasOpenBracket && token === '{}') isValid = false
+    // after a closing bracket, only an opening bracket or {} are allowed
+    if (token === ']' && nextToken && !['[', '{}'].includes(nextToken)) {
+      isValid = false
+      break
+    }
+  }
+
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
```

Comment:
Your examples are `form.append('foo{}', '{"bar":[1,2,3]}') `and more complex keys. But what if the value is a primitive, like 'a' or worse 'null'?

secureParse would not fail in case of null.

Reply:
I've included an explicit check for `null`, but `secureJSON.parse` would, in fact, allow primitives to be passed. Is there a good reason for explicitly preventing this?